pkg/kubelet/util/pluginwatcher/example_handler.go

```diff
@@ -0,0 +1,107 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pluginwatcher
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"golang.org/x/net/context"
+
+	v1beta1 "k8s.io/kubernetes/pkg/kubelet/util/pluginwatcher/example_plugin_apis/v1beta1"
+	v1beta2 "k8s.io/kubernetes/pkg/kubelet/util/pluginwatcher/example_plugin_apis/v1beta2"
+)
+
+type exampleHandler struct {
+	registeredPlugins        map[string]bool
+	mutex                    sync.Mutex
+	chanForHandlerAckTesting chan error // for testing
+}
+
+func NewExampleHandler() *exampleHandler {
+	return &exampleHandler{
+		chanForHandlerAckTesting: make(chan error),
+		registeredPlugins:        make(map[string]bool),
+	}
+}
+
+func (h *exampleHandler) Cleanup() error {
+	h.mutex.Lock()
+	defer h.mutex.Unlock()
+	h.registeredPlugins = make(map[string]bool)
+	return nil
+}
+
+func (h *exampleHandler) Handler(pluginName string, endpoint string, versions []string, sockPath string) (error, chan bool) {
+
+	// check for supported versions
+	if equal := reflect.DeepEqual([]string{"v1beta1", "v1beta2"}, versions); !equal {
+		return fmt.Errorf("not a supported version"), nil
+	}
+
+	// this handler expect non-empty endpoint as a example
+	if len(endpoint) == 0 {
+		return fmt.Errorf("endpoint can not be empty"), nil
+	}
+
+	_, conn, err := dial(sockPath)
+	if err != nil {
+		return err, nil
+	}
+	defer conn.Close()
+
+	// The plugin handler should be able to use any listed service API version.
+	v1beta1Client := v1beta1.NewExampleClient(conn)
+	v1beta2Client := v1beta2.NewExampleClient(conn)
+
+	// Tests v1beta1 GetExampleInfo
+	_, err = v1beta1Client.GetExampleInfo(context.Background(), &v1beta1.ExampleRequest{})
+	if err != nil {
+		return err, nil
+	}
+
+	// Tests v1beta1 GetExampleInfo
+	_, err = v1beta2Client.GetExampleInfo(context.Background(), &v1beta2.ExampleRequest{})
+	if err != nil {
+		return err, nil
+	}
+
+	// handle registered plugin
+	h.mutex.Lock()
```

Comment:
Hi @figo,

Please remove lines 89 and 93 - `h.mutex.Unlock()` and relocate that to beneath the `Lock()` call as: `defer h.mutex.Unlock()`. This ensures the mutex is released no matter where the function is returned.

Reply:
let me try to explain why `defer` is not suitable here,  we want to release lock as soon as we can, which means right after line 92, otherwise, has to wait until the end of function.

Reply:
I understand that, but you’ve got a timer that guarantees the function takes no longer than one second anyway, so is it really that much of hit to ensure no possible lock error occurs?

Reply:
Hi @figo,

Except efficiency isn't always required when a code path isn't part of the happy path. It looks here like you're just checking to see if a plug-in has been registered, and marking it as such if it has not. Is that really executed so often as to sacrifice design for squeaking out the smallest performance gains?

Additionally:

1. The lock should be a `sync.RWMutex` regardless as read-locks are less expensive, and a map is a classic case of using both read and write locks.
2. The code should be rewritten:

```go
	h.mutex.RLock()
	_, registered := h.registeredPlugins[pluginName]
	h.mutex.RUnlock()
	if registered {
 		return nil, fmt.Errorf("duplicate plugin registration: %s", pluginName)
 	}
 	h.registeredPlugins[pluginName] = true
```

Because the mutex isn't used to synchronize the entire function, there's no reason to even hold the lock beyond checking to see if the plug-in has been registered.

Reply:
Hi @figo,

Please disregard part of the above statement. I'm rewriting it now. Because of the diff I missed part of the lock scope.

Reply:
Hi @figo,

I missed the registration inside the scope of the lock. Your design makes sense, but I also wonder if this isn't the case of a `sync.Map`? I would also consider creating a helper function to make things neater if you want to stick with a stdlib `map`:

```go
func (h *Handler) registerPlugin(name string) error {
	h.mutex.Lock()
	defer h.mutex.Unlock()
	if _, ok := h.registeredPlugins[pluginName]; !ok {
		return nil, fmt.Errorf("duplicate plugin registration: %s", name)
	}
	 h.registeredPlugins[name] = true
}
```